File: bridge/EVSimulatorApp.Bridge.Tests/SessionConfigTests.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;
using NUnit.Framework;

namespace EVSimulatorApp.Bridge.Tests;

/// <summary>
/// The one thing that travels <em>into</em> the bridge, and the corpus the ports are held to.
/// </summary>
/// <remarks>
/// <para>
/// The event stream is a record: a page can only watch it. This is the opposite direction — a value
/// a WebView hands to native code, which then opens a socket because of it. So the interesting half
/// of this file is the refusals, and the refusals are pinned <b>by message</b>: three back ends that
/// all say no for three different reasons are three different products, and the difference only ever
/// shows up in front of a user who cannot act on it.
/// </para>
/// <para>
/// <b>Nothing here resolves a host name</b>, by the same rule as the pairing sheet: a lookup on
/// behalf of a code nobody has agreed to trust is already a callback to whoever printed it.
/// </para>
/// </remarks>
[TestFixture]
public class SessionConfigTests
{

    private const string CorpusFile = "Bridge.config.json";

    private static readonly JsonSerializerOptions Pretty = new() { WriteIndented = true };


    private static string RepositoryRoot
    {
        get
        {
            var directory = new DirectoryInfo(TestContext.CurrentContext.TestDirectory);
            while (directory is not null && !Directory.Exists(Path.Combine(directory.FullName, "libs/Vanaheimr.V2G.Exi")))
                directory = directory.Parent;

            return directory?.FullName ?? throw new DirectoryNotFoundException("repository root not found");
        }
    }

    private static string CorpusPath =>
        Path.Combine(RepositoryRoot, "bridge/EVSimulatorApp.Bridge.Tests/Vectors", CorpusFile);


    #region The cases

    /// <summary>Configurations a session may start from.</summary>
    private static readonly (string Name, string Json)[] Accepted = [

        ("minimal", """
            {"host":"192.168.1.42","port":15118,"transport":"tls","protocol":"iso15118-2","mode":"ac","authorization":"eim"}
            """),

        ("everything", """
            {"host":"10.1.2.3","port":15119,"transport":"tcp","protocol":"iso15118-20","mode":"dc",
             "authorization":"pnc","totp":"481923",
             "rootFingerprint":"9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08"}
            """),

        // A name, never an address. `.local` is the only name shape that counts as reachable-but-local,
        // and it is judged as text — see PairingWarnings.IsPrivateTarget.
        ("mdns-name", """
            {"host":"secc-1.local","port":15118,"transport":"tls","protocol":"iso15118-2","mode":"ac","authorization":"eim"}
            """),

        // ISO 15118 runs over link-local IPv6 in the wired case, zone identifier and all.
        ("ipv6-link-local", """
            {"host":"fe80::1%en0","port":15118,"transport":"tls","protocol":"iso15118-20","mode":"dc","authorization":"eim"}
            """),

        ("ipv6-unique-local", """
            {"host":"fd12:3456::1","port":15118,"transport":"tls","protocol":"iso15118-20","mode":"ac","authorization":"pnc"}
            """),

        ("loopback", """
            {"host":"127.0.0.1","port":1,"transport":"tcp","protocol":"iso15118-2","mode":"ac","authorization":"eim"}
            """),

        ("highest-port", """
            {"host":"172.16.0.1","port":65535,"transport":"tls","protocol":"iso15118-2","mode":"dc","authorization":"eim"}
            """),

        // Absent and explicitly null mean the same thing, because the writer omits rather than nulls
        // and a null can therefore only come from somebody else's writer.
        ("explicit-nulls", """
            {"host":"169.254.10.1","port":15118,"transport":"tls","protocol":"iso15118-2","mode":"ac",
             "authorization":"eim","totp":null,"rootFingerprint":null}
            """),

    ];


    /// <summary>Configurations no session starts from, and the words the user is shown.</summary>
    private static readonly (string Name, string Json)[] Refused = [

        ("not-an-object",        """ ["192.168.1.42", 15118] """),
        ("null-document",        """ null """),

        // A key this build does not read is either a newer front end or somebody probing. Ignoring it
        // would mean a setting shown on the confirmation sheet silently did not happen.
        ("unknown-property",     """
            {"host":"192.168.1.42","port":15118,"transport":"tls","protocol":"iso15118-2","mode":"ac",
             "authorization":"eim","tlsVersion":"1.2"}
            """),

        // B1's private-range restriction, applied where the socket is opened rather than where the
        // sheet was drawn.
        ("public-address",       """
            {"host":"93.184.216.34","port":15118,"transport":"tls","protocol":"iso15118-2","mode":"ac","authorization":"eim"}
            """),

        ("resolvable-name",      """
            {"host":"station.example.com","port":15118,"transport":"tls","protocol":"iso15118-2","mode":"ac","authorization":"eim"}
            """),

        // 010.0.0.1 is 8.0.0.1 to a C resolver and 10.0.0.1 to a reader in a hurry. Neither reading
        // is used: a leading zero makes it not an address literal at all, so it is judged as a name
        // and refused for not ending in .local.
        ("octal-looking-host",   """
            {"host":"010.0.0.1","port":15118,"transport":"tls","protocol":"iso15118-2","mode":"ac","authorization":"eim"}
            """),

        ("public-ipv6",          """
            {"host":"2001:db8::1","port":15118,"transport":"tls","protocol":"iso15118-2","mode":"ac","authorization":"eim"}
            """),

        ("missing-host",         """
            {"port":15118,"transport":"tls","protocol":"iso15118-2","mode":"ac","authorization":"eim"}
            """),

        ("empty-host",           """
            {"host":"","port":15118,"transport":"tls","protocol":"iso15118-2","mode":"ac","authorization":"eim"}
            """),

        ("host-not-a-string",    """
            {"host":42,"port":15118,"transport":"tls","protocol":"iso15118-2","mode":"ac","authorization":"eim"}
            """),

        ("missing-port",         """
            {"host":"192.168.1.42","transport":"tls","protocol":"iso15118-2","mode":"ac","authorization":"eim"}
            """),

        ("port-zero",            """
            {"host":"192.168.1.42","port":0,"transport":"tls","protocol":"iso15118-2","mode":"ac","authorization":"eim"}
            """),

        ("port-above-range",     """
            {"host":"192.168.1.42","port":65536,"transport":"tls","protocol":"iso15118-2","mode":"ac","authorization":"eim"}
            """),

        ("port-as-string",       """
            {"host":"192.168.1.42","port":"15118","transport":"tls","protocol":"iso15118-2","mode":"ac","authorization":"eim"}
            """),

        ("port-fractional",      """
            {"host":"192.168.1.42","port":15118.5,"transport":"tls","protocol":"iso15118-2","mode":"ac","authorization":"eim"}
            """),

        ("unknown-transport",    """
            {"host":"192.168.1.42","port":15118,"transport":"quic","protocol":"iso15118-2","mode":"ac","authorization":"eim"}
            """),

        // Exact match, not case-insensitive. The producer is a program, not a person.
        ("transport-uppercased", """
            {"host":"192.168.1.42","port":15118,"transport":"TLS","protocol":"iso15118-2","mode":"ac","authorization":"eim"}
            """),

        ("unknown-protocol",     """
            {"host":"192.168.1.42","port":15118,"transport":"tls","protocol":"din70121","mode":"ac","authorization":"eim"}
            """),

        ("unknown-mode",         """
            {"host":"192.168.1.42","port":15118,"transport":"tls","protocol":"iso15118-20","mode":"wpt","authorization":"eim"}
            """),

        ("unknown-authorization","""
            {"host":"192.168.1.42","port":15118,"transport":"tls","protocol":"iso15118-2","mode":"ac","authorization":"none"}
            """),

        // An empty optional is a value somebody meant to supply and did not — unlike an absent one.
        ("empty-totp",           """
            {"host":"192.168.1.42","port":15118,"transport":"tls","protocol":"iso15118-2","mode":"ac",
             "authorization":"eim","totp":""}
            """),

        ("totp-not-a-string",    """
            {"host":"192.168.1.42","port":15118,"transport":"tls","protocol":"iso15118-2","mode":"ac",
             "authorization":"eim","totp":481923}
            """),

    ];

    #endregion


    /// <summary>The corpus, as this back end produces it.</summary>
    private static JsonObject Produce()
    {

        var accepted = new JsonArray();

        foreach (var (name, text) in Accepted)
        {
            var config = SessionConfig.Parse(JsonNode.Parse(text));

            accepted.Add((JsonNode?) new JsonObject {
                ["name"]      = name,
                ["input"]     = JsonNode.Parse(text),
                ["canonical"] = config.ToJSON(),
            });
        }

        var refused = new JsonArray();

        foreach (var (name, text) in Refused)
        {
            string message;
            try
            {
                SessionConfig.Parse(JsonNode.Parse(text));
                message = "(accepted — this case no longer refuses anything)";
            }
            catch (SessionConfigException e)
            {
                message = e.Message;
            }

            refused.Add((JsonNode?) new JsonObject {
                ["name"]    = name,
                ["input"]   = JsonNode.Parse(text),
                ["message"] = message,
            });
        }

        return new JsonObject {
            ["note"]     = "Generated by SessionConfigTests.RegenerateTheCorpus. What a WebView may ask "
                         + "the native side to connect to, and — the larger half — what it may not. "
                         + "The refusal messages are part of the corpus: a port that says no for a "
                         + "different reason is a different product.",
            ["accepted"] = accepted,
            ["refused"]  = refused,
        };
    }


    [Test, Explicit("Rewrites the checked-in configuration corpus. Run deliberately, and read the diff.")]
    public void RegenerateTheCorpus()
    {
        Directory.CreateDirectory(Path.GetDirectoryName(CorpusPath)!);
        File.WriteAllText(CorpusPath, Produce().ToJsonString(Pretty));

        TestContext.Out.WriteLine($"-> {CorpusPath}");
    }


    [Test]
    public void TheCorpusIsUnchanged()
    {

        Assert.That(File.Exists(CorpusPath), Is.True, $"the configuration corpus is missing at {CorpusPath}");

        var expected = JsonNode.Parse(File.ReadAllText(CorpusPath))!.AsObject();
        var actual   = Produce();

        foreach (var section in new[] { "accepted", "refused" })
            Assert.That(actual[section]!.ToJsonString(Pretty),
                        Is.EqualTo(expected[section]!.ToJsonString(Pretty)), section);
    }


    /// <summary>
    /// Every accepted configuration survives its own JSON: parse, write, parse again, same value.
    /// </summary>
    /// <remarks>
    /// The round trip is what makes the canonical form usable as an audit record. It is deliberately
    /// <em>not</em> the only check here — a round trip is blind to what the properties are called,
    /// because the writer and the reader rename together. That is what the corpus above pins, and
    /// what <see cref="TheTypeScriptSideDeclaresTheSameProperties"/> holds to the front end.
    /// </remarks>
    [Test]
    public void EveryAcceptedConfigurationSurvivesItsOwnJson()
    {

        foreach (var (name, text) in Accepted)
        {
            var once  = SessionConfig.Parse(JsonNode.Parse(text));
            var twice = SessionConfig.Parse(once.ToJSON());

            Assert.That(twice, Is.EqualTo(once), name);
        }
    }


    /// <summary>
    /// The properties this reads are the properties the TypeScript front end writes.
    /// </summary>
    /// <remarks>
    /// <para>
    /// The producer and the consumer of this shape live in different languages, in different
    /// processes, and are never compiled together — which is precisely the arrangement in which a
    /// property quietly stops being read. Since unknown properties are <em>refused</em>, a front end
    /// that adds one does not degrade gracefully: it fails to start a session at all. So the drift is
    /// worth catching here rather than on a phone.
    /// </para>
    /// <para>
    /// Read out of the TypeScript source rather than duplicated, because a second list is the thing
    /// that would go stale.
    /// </para>
    /// </remarks>
    [Test]
    public void TheTypeScriptSideDeclaresTheSameProperties()
    {

        var source = File.ReadAllText(Path.Combine(RepositoryRoot, "typescript/src/bridge/plugin.ts"));
        var body   = Regex.Match(source, @"export interface SessionConfig \{(.*?)\n\}", RegexOptions.Singleline);

        Assert.That(body.Success, Is.True, "the SessionConfig interface was not found in plugin.ts");

        var declared = Regex.Matches(body.Groups[1].Value, @"readonly\s+([A-Za-z0-9_]+)\??\s*:")
                            .Select(m => m.Groups[1].Value)
                            .ToList();

        var known = SessionConfig.Parse(JsonNode.Parse(Accepted[1].Json)).ToJSON().Select(p => p.Key).ToList();

        Assert.That(declared, Is.EquivalentTo(known),
                    "typescript/src/bridge/plugin.ts and SessionConfig.cs no longer describe the same value");
    }

}

File: swift/Tests/V2GBridgeTests/SessionConfigTests.swift
import XCTest
import Foundation
import ExiRuntime
@testable import V2GBridge

/// This back end's reading of a session configuration, against C#'s.
///
/// The event stream is a record: a page can only watch it. This is the opposite direction — a value a
/// WebView hands to native code, which then opens a socket because of it. So the interesting half is
/// the refusals, and they are pinned **by message**: a port that says no for a different reason is a
/// different product, and the difference only ever shows up in front of a user who cannot act on it.
final class SessionConfigTests: XCTestCase {

    private lazy var repositoryRoot: URL = {
        var directory = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
        while !FileManager.default.fileExists(
                atPath: directory.appendingPathComponent("EVSimulatorApp.slnx").path) {
            let parent = directory.deletingLastPathComponent()
            precondition(parent != directory, "repository root not found")
            directory = parent
        }
        return directory
    }()

    private lazy var corpus: JsonObject = {
        let file = repositoryRoot
            .appendingPathComponent("bridge/EVSimulatorApp.Bridge.Tests/Vectors/Bridge.config.json")
        let text = try! String(contentsOf: file, encoding: .utf8)
        return try! JsonValue.parse(text) as! JsonObject
    }()

    private func section(_ name: String) -> [JsonObject] {
        (corpus[name] as! JsonArray).asList.map { $0 as! JsonObject }
    }

    private func caseName(_ entry: JsonObject) -> String {
        (entry["name"] as! JsonString).value
    }


    func testEveryConfigurationCSharpAcceptsIsAcceptedHereCanonically() throws {

        let cases = section("accepted")
        XCTAssertGreaterThanOrEqual(cases.count, 8, "only \(cases.count) accepted cases")

        for entry in cases {
            let config = try SessionConfig.parse(entry["input"])

            XCTAssertEqual(config.toJSON().jsonString,
                           (entry["canonical"] as! JsonObject).jsonString,
                           caseName(entry))
        }
    }


    func testEveryConfigurationCSharpRefusesIsRefusedHereInTheSameWords() throws {

        let cases = section("refused")
        XCTAssertGreaterThanOrEqual(cases.count, 20, "only \(cases.count) refused cases")

        for entry in cases {

            let expected = (entry["message"] as! JsonString).value

            do {
                _ = try SessionConfig.parse(entry["input"])
                XCTFail("\(caseName(entry)) was accepted; C# refuses it with: \(expected)")
            } catch let error as SessionConfigError {
                XCTAssertEqual(error.description, expected, caseName(entry))
            }
        }
    }


    /// Parse, write, parse again — the same value. What makes the canonical form usable as a record.
    func testEveryAcceptedConfigurationSurvivesItsOwnJson() throws {

        for entry in section("accepted") {
            let once  = try SessionConfig.parse(entry["input"])
            let twice = try SessionConfig.parse(once.toJSON())

            XCTAssertEqual(once, twice, caseName(entry))
        }
    }


    /// A runner that cannot start refuses the command; it does not open a stream that immediately dies.
    ///
    /// The distinction is what a caller acts on: a throw means no session exists and the WebView's
    /// `start()` rejects, while an error event means a session is running and something in it failed.
    func testATraceRunnerWithNoRecordingRefusesRatherThanEmittingABrokenStream() throws {

        let runner = TraceSessionRunner(trace: { _ in nil }, monotonicMillis: SteppingClock().read)
        let config = try SessionConfig.parse(section("accepted").first!["input"])

        var emitted: [BridgeEvent] = []

        XCTAssertThrowsError(try runner.run(config) { emitted.append($0) }) { error in
            XCTAssertTrue((error as! SessionRunnerError).description.contains(config.protocol),
                          "the refusal names what was asked for")
        }
        XCTAssertTrue(emitted.isEmpty, "a refused command emitted \(emitted.count) event(s)")
    }


    /// The measurement the Capacitor payload's shape rests on: an event does not survive
    /// `JSONSerialization`.
    ///
    /// Capacitor's `notifyListeners` takes a `[String: Any]`, and `JSONSerialization` is the only way
    /// to build one from a document. It is backed by a `Dictionary`, which has no member order at
    /// all — so every JSON-LD document that went through it would arrive in an order decided by a
    /// hash function, differing from Android's and from the corpus every back end agrees on. That is
    /// why an event crosses the bridge as **text**.
    ///
    /// Kept here rather than in the plugin because the plugin can only be built for iOS, and a
    /// measurement nothing runs is a comment. If this ever stops failing — if some future
    /// `JSONSerialization` preserved order — the text payload would become ceremony rather than
    /// protection, and this test is where that would be noticed.
    func testAnEventDoesNotSurviveJSONSerialization() throws {

        let events = try JsonValue.parse(String(contentsOf: repositoryRoot.appendingPathComponent(
            "bridge/EVSimulatorApp.Bridge.Tests/Vectors/Bridge.events.json"), encoding: .utf8))

        let sessions = (events as! JsonObject)["sessions"] as! JsonObject

        var total = 0, changed = 0

        for name in sessions.keys {
            for event in (sessions[name] as! JsonArray).asList {

                let text   = event.jsonString
                let object = try JSONSerialization.jsonObject(with: Data(text.utf8))
                let data   = try JSONSerialization.data(withJSONObject: object,
                                                        options: [.withoutEscapingSlashes])
                total += 1
                if String(decoding: data, as: UTF8.self) != text { changed += 1 }
            }
        }

        XCTAssertGreaterThanOrEqual(total, 190, "only \(total) events were measured")
        XCTAssertEqual(changed, total,
            "JSONSerialization preserved \(total - changed) of \(total) events — re-read BridgeCodec")
    }


    /// And one that has a recording produces exactly the stream the event corpus pins.
    func testATraceRunnerReplaysTheRecordedSession() throws {

        let traceDirectory = repositoryRoot
            .appendingPathComponent("../ISO15118ConformanceTests.Simulation/Vectors")

        let traceFile = try FileManager.default
            .contentsOfDirectory(atPath: traceDirectory.path)
            .filter { $0.hasPrefix("Session.") && $0.hasSuffix(".trace.json") }
            .sorted()
            .first!

        let trace = try JsonValue.parse(
            String(contentsOf: traceDirectory.appendingPathComponent(traceFile), encoding: .utf8))
            as! JsonObject

        let name   = (trace["name"] as! JsonString).value
        let runner = TraceSessionRunner(trace: { _ in trace }, monotonicMillis: SteppingClock().read)
        let config = try SessionConfig.parse(section("accepted").first!["input"])

        var emitted: [BridgeEvent] = []
        try runner.run(config) { emitted.append($0) }

        let events = try JsonValue.parse(String(contentsOf: repositoryRoot.appendingPathComponent(
            "bridge/EVSimulatorApp.Bridge.Tests/Vectors/Bridge.events.json"), encoding: .utf8))
        let expected = ((events as! JsonObject)["sessions"] as! JsonObject)[name] as! JsonArray

        XCTAssertEqual(emitted.count, expected.count, "\(name): event count")

        for (i, event) in emitted.enumerated() {
            XCTAssertEqual(BridgeEvent.toJSON(event).jsonString, expected.asList[i].jsonString,
                           "\(name)/\(i)")
        }
    }

}
